### PR TITLE
[omdb] add basic support for activating background tasks

### PR DESCRIPTION
**Pull request by sunshowers**
This does its basic job of activating a background task, e.g.
`inventory_collection`.

It is a little unsatisfying because it's a bit hard with the current structure
to do things like:

* if a task is currently running, wait until that is done
* wait until the task we triggered is finished
* return some kind of identifier (e.g. collection ID for inventory_collection)

Providing this kind of progress reporting is one of the kinds of problems we
solved with the update engine, and I'm wondering if it makes sense to try and
integrate that at some point.

Fixes https://github.com/oxidecomputer/omicron/issues/5058.

### CODE REVIEW

File: nexus/src/internal_api/http_entrypoints.rs
Comment (jgallagher) [inline nexus/src/internal_api/http_entrypoints.rs:788]: Awkward! I guess this is because as far as dropshot / OpenAPI is concerned, we could have a background task _named_ `activate`? Would it make sense to move the GET endpoints down a level? `/bgtasks/status/{bgtask_name}` or something like that

Comment (davepacheco) [reply nexus/src/internal_api/http_entrypoints.rs:788]: It's dropshot that enforces that you can't have static paths and variable paths that overlap.  It obviously could be made to work, but seemed like the sort of thing we want to avoid.  I'd be fine with moving that down to `/bgtasks/status`.  It'll be a runtime flag day between `omdb` and Nexus but I don't imagine a lot of people use mismatched `omdb`s and invoke these commands a lot.

File: dev-tools/omdb/src/bin/omdb/nexus.rs
Comment (sunshowers) [reply dev-tools/omdb/src/bin/omdb/nexus.rs:484]: I was wondering about that, it didn't feel like a destructive operation?

Comment (davepacheco) [reply dev-tools/omdb/src/bin/omdb/nexus.rs:484]: I would add it.  This terminology and flag came from `mdb` and `dtrace` where "destructive" means "may alter the thing you're debugging".  That might not destroy anything per se but I think it is the most useful distinction here (i.e., read-only vs. not).

File: nexus/src/internal_api/http_entrypoints.rs
Comment (sunshowers) [reply nexus/src/internal_api/http_entrypoints.rs:788]: Yeah. Completely agree that we want to avoid conflicts here -- I'll change the API to `/bgtasks/status` etc, and just assume that the omdb flag day isn't a huge deal.

**Comment by sunshowers** [comment]
> I suspect that what most programmatic consumers want is not to wait for a specific activation to complete but rather for some specific set of changes to be applied, which could take many laps. (Complicating it further, they might also want to give up waiting if there was a successful lap and the change was not applied.)

I think I generally agree with this model for programmatic consumers of background tasks (especially as compared to wicketd, which is much more of a foreground/oneshot approach).

I think humans would benefit greatly from being able to wait until the next activation after the trigger point completes. As a human looking at a live system, it can be hard to express exactly the constraint one is waiting for. (You could write Rust code for specific cases, like waiting for a new sled to show up in inventory for the first time, but in general I think you'd need some kind of dynamic query language that covers everything operators could reasonable ask for.)

In general, I think transient errors like not being able to talk to a sled will be quite rare -- and if an operator or support tech is actively monitoring the situation, they can observe those errors and kick off another run of the background task.

The update engine can also be used for post-mortem analysis as long as the generated events are written to a log file -- we already have code that can read and replay logs in a nice fashion.

**Comment by davepacheco** [comment]
> I think humans would benefit greatly from being able to wait until the next activation after the trigger point completes. As a human looking at a live system, it can be hard to express exactly the constraint one is waiting for. (You could write Rust code for specific cases, like waiting for a new sled to show up in inventory for the first time, but in general I think you'd need some kind of dynamic query language that covers everything operators could reasonable ask for.)

Is this something where we could leave it to the human to poll by hand on whatever condition they care about (e.g., `watch --chgexit omdb db inventory list`)?

I also think it would be fine to have richer debugging for tasks like you're suggesting.  We could have APIs to list recent activations and their statuses and/or to wait for an activation to complete.  Or to wait for activation N to complete.  I just think there's a ton of stuff we _could_ do here (so there's some risk of scope creep) and some of it could be easily misused to build brittle stuff.  So I'm wondering how big a pain point it currently is.  If it is painful today in dev then yeah maybe we should do it and see if we can make sure people don't accidentally use it to build programmatic consumers that ought to be checking other conditions instead.

File: dev-tools/omdb/src/bin/omdb/nexus.rs
Comment (sunshowers) [reply dev-tools/omdb/src/bin/omdb/nexus.rs:484]: Ah gotcha, thanks!

**Comment by sunshowers** [comment]
Yeah, let's see how it plays out over the next while as we add more functionality to our background task system, and decide from there.

**Review by andrewjstone — APPROVED** [review]

**Comment by sunshowers** [comment]
Need to wait until 5620 lands.

**Pull request merged**
